docs/development/secure-development-process.md

```diff
@@ -0,0 +1,202 @@
+
+**Avoid complex regexes**
+Complex regexes can lead to potential denial of service attacks. This article provides a good overview of the risks: [Regular_expression_Denial_of_Service_-_ReDoS](https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS)
+
+**Limit the attack surface**
+
+Limit the available attack surface by:
+* Only exposing APIs that are needed to support the intended
+  functionality. For example, when using express remove any uncessary routes.
+* Group all external endpoints under a prefix(for example /api). This
+  makes it easier to only expose APIs which are intended to
+  be external in the ingress configuration. Don’t path rewrite to /.
+* Limiting access based on authentication. When possible integrate
+  an organizational Identity and access control provider instead of
+  implementation your own.
+
+**Harden external HTTP endpoints**
+
+Harden external http endpoints by:
+* Adding defensive HTTP headers using packages like [Helmet](https://www.npmjs.com/package/helmet).
+* Taking precautions against brute force and denial of service 
+  attacks. These can come in different types. Most often the team uses
+  the following to address each type:
+  * Denial of service -> Content delivery network (CDN) and rate limit login
+  * One customer using too much -> External API manager (for example Red Hat 3scale API management)
+  * Over sized requests -> Enforcing request size limits
+* Prevent HTTP parameter polution with packages like
+  [hpp](https://www.npmjs.com/package/hpp)
+* Protect against Cross site scripting (XSS). For example by using a package like [xss](https://www.npmjs.com/package/xss).
+* Protect against cross site forgery requests
+  * Use Anti-CSRF tokens through packages like [csurf](https://www.npmjs.com/package/csurf)
+
+**Avoid leaking info through errors**
+
+Do not expose senstive or intermal information in errors returned to the end user. Instead return a unique ID that can be tied back to more specific error information that is written to internals logs.
+
+**Limit data sent back to front ends**
+
+Only send the required data to the front end. For example if a database query returns 4 fields but the front end will only display 3, make sure to remove the fourth field before sending the data on to the front end.
+
+**Discard sentive data after use**
+
+Do not retain senstive information beyond where it is needed. Overwrite fields/variables that contain sensitive data once it is not longer needed.
+
+**Validate input**
+
+Unvalidated input can result in:
+* command injection
+* sql injection
+* data corruption
+* denial of service
+
+Do not pass unsanitized user input to the Node.js [child_process](https://nodejs.org/api/child_process.html) functions as this can result in command injection where an attacker can run a common on the server on which Node.js is running. This is a good article that goes into more details: 
```

Comment:
```suggestion
Do not pass unsanitized user input to the Node.js [child_process](https://nodejs.org/api/child_process.html) functions as this can result in command injection where an attacker can run a command on the server on which Node.js is running. This is a good article that goes into more details: 
```